src/platform/plugins/shared/dashboard/public/dashboard_container/component/grid/dashboard_grid.tsx

```diff
@@ -7,153 +7,134 @@
+      if (viewMode !== 'edit') return;
+
+      const currentPanels = dashboardApi.panels$.getValue();
+      const updatedPanels: { [key: string]: DashboardPanelState } = Object.values(
+        newLayout[0].panels
+      ).reduce((updatedPanelsAcc, panelLayout) => {
+        updatedPanelsAcc[panelLayout.id] = {
+          ...currentPanels[panelLayout.id],
+          gridData: {
+            i: panelLayout.id,
+            y: panelLayout.row,
+            x: panelLayout.column,
+            w: panelLayout.width,
+            h: panelLayout.height,
+          },
+        };
+        return updatedPanelsAcc;
+      }, {} as { [key: string]: DashboardPanelState });
+      if (!arePanelLayoutsEqual(currentPanels, updatedPanels)) {
+        dashboardApi.setPanels(updatedPanels);
+      }
+    },
+    [dashboardApi, viewMode]
+  );
+
+  const renderPanelContents = useCallback(
+    (id: string, setDragHandles?: (refs: Array<HTMLElement | null>) => void) => {
+      const currentPanels = dashboardApi.panels$.getValue();
+      if (!currentPanels[id]) return;
+
+      if (!panelRefs.current[id]) {
+        panelRefs.current[id] = React.createRef();
+      }
+
+      const type = currentPanels[id].type;
       return (
         <DashboardGridItem
+          ref={panelRefs.current[id]}
+          data-grid={currentPanels[id].gridData}
+          key={id}
+          id={id}
+          type={type}
+          setDragHandles={setDragHandles}
           appFixedViewport={appFixedViewport}
           dashboardContainer={dashboardContainer}
-          data-grid={panels[embeddableId].gridData}
-          key={embeddableId}
-          id={embeddableId}
-          index={index + 1}
-          type={type}
-          expandedPanelId={expandedPanelId}
-          focusedPanelId={focusedPanelId}
         />
       );
-    });
-  }, [
-    appFixedViewport,
-    dashboardContainer,
-    expandedPanelId,
-    panels,
-    panelsInOrder,
-    focusedPanelId,
-  ]);
-
-  const onLayoutChange = useCallback(
-    (newLayout: Array<Layout & { i: string }>) => {
-      if (viewMode !== 'edit') return;
-
-      const updatedPanels: { [key: string]: DashboardPanelState } = newLayout.reduce(
-        (updatedPanelsAcc, panelLayout) => {
-          updatedPanelsAcc[panelLayout.i] = {
-            ...panels[panelLayout.i],
-            gridData: pick(panelLayout, ['x', 'y', 'w', 'h', 'i']),
-          };
-          return updatedPanelsAcc;
-        },
-        {} as { [key: string]: DashboardPanelState }
-      );
-      if (!arePanelLayoutsEqual(panels, updatedPanels)) {
-        dashboardApi.setPanels(updatedPanels);
-      }
     },
-    [dashboardApi, panels, viewMode]
+    [appFixedViewport, dashboardApi, dashboardContainer]
   );
 
+  const memoizedgridLayout = useMemo(() => {
+    // memoizing this component reduces the number of times it gets re-rendered to a minimum
+    return (
+      <GridLayout
+        layout={currentLayout}
+        gridSettings={{
```

Comment:
This looks really clean with props instead of a `getCreationOptions` callback. Really nice work.